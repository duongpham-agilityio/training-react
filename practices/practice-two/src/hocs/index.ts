export * from './withErrorBoundaries';
export * from './withUseForm';
