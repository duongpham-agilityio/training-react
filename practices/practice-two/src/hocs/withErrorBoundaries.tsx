import React, { FunctionComponent, Component } from 'react';

// Component
import { Error } from 'components';

export const withErrorBoundaries = <T extends object>(
  WrappedComponent: FunctionComponent<T>
) => {
  class Wrapper extends Component<T, { hashError: boolean }> {
    constructor(props: T) {
      super(props);
      this.state = {
        hashError: false,
      };
    }

    componentDidCatch(): void {
      this.setState({
        hashError: true,
      });
    }

    render(): React.ReactNode {
      if (this.state.hashError) {
        return <Error />;
      }

      return <WrappedComponent {...this.props} />;
    }
  }

  return Wrapper;
};
