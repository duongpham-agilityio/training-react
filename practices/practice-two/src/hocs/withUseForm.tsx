import { FC, Suspense, lazy, useCallback, useMemo, useState } from 'react';
import useMutation from 'swr/mutation';

// Hooks
import { FormData, useToastContext } from 'hooks';

// Components
const Form = lazy(() => import('components/Form'));

// Constants
import { ENDPOINT, MESSAGES } from '@constants';

// Helpers
import { axiosConfig } from 'helpers';

// Types
import { Book } from 'types';

type FormStatus = 'update' | 'create';
interface StateType {
  formData: Omit<Book, 'publishDate'> & {
    publishDate?: number;
  };
  title: string;
  type: FormStatus;
}

export interface WithUseFormProps {
  dispatchAction: (data: StateType) => void;
}

export const withUseForm = <T extends WithUseFormProps>(Component: FC<T>) => {
  const FormHOC = (props: Omit<T, keyof WithUseFormProps>) => {
    const { setNotification } = useToastContext();
    const [form, setForm] = useState<StateType | null>(null);
    const [formStatus, setFormStatus] = useState<FormStatus>('create');
    const { trigger: triggerAddBook } = useMutation(
      `${ENDPOINT.BOOKS}${ENDPOINT.SORT}`,
      (_, { arg }: { arg: Omit<FormData, 'imageName'> }) =>
        axiosConfig.post(ENDPOINT.BOOKS, arg)
    );
    const { trigger: triggerUpdateBook } = useMutation(
      `${ENDPOINT.BOOKS}/${form?.formData.id}`,
      (key: string, { arg }: { arg: FormData & { updatedAt: number } }) =>
        axiosConfig.patch(key, arg)
    );

    const handleCloseForm = useCallback(() => setForm(null), []);

    const dispatchAction = useCallback((data: StateType) => {
      setForm(data);
      setFormStatus(data.type);
    }, []);

    const onTriggerCreate = useCallback(
      (result: FormData) =>
        triggerAddBook(result)
          .then(() =>
            setNotification({
              message: MESSAGES.ADD_SUCCESS,
              title: MESSAGES.ADD_TITLE,
            })
          )
          .catch(() =>
            setNotification({
              message: MESSAGES.ERROR_TITLE,
              title: MESSAGES.EMPTY_FIELD,
              type: 'error',
            })
          )
          .finally(handleCloseForm),
      [handleCloseForm, setNotification, triggerAddBook]
    );

    const onTriggerUpdate = useCallback(
      (result: FormData) => {
        const { author, description, imageURL, name, publishDate } = result;
        const payload: FormData & { updatedAt: number } = {
          author,
          description,
          imageURL,
          name,
          updatedAt: new Date().getTime(),
          publishDate,
        };

        return triggerUpdateBook(payload)
          .then(() =>
            setNotification({
              message: MESSAGES.UPDATE_TITLE,
              title: MESSAGES.UPDATE_SUCCESS,
            })
          )
          .catch(() =>
            setNotification({
              message: MESSAGES.ERROR_TITLE,
              title: MESSAGES.EMPTY_FIELD,
              type: 'error',
            })
          )
          .finally(handleCloseForm);
      },
      [handleCloseForm, setNotification, triggerUpdateBook]
    );

    const onSubmit = useMemo(() => {
      const handler = {
        create: onTriggerCreate,
        update: onTriggerUpdate,
      };

      return handler[formStatus];
    }, [formStatus, onTriggerCreate, onTriggerUpdate]);

    return (
      <>
        <Component {...(props as T)} {...{ dispatchAction }} />
        {form && (
          <Suspense>
            <Form
              value={form.formData}
              title={form.title}
              type={form.type}
              onClose={handleCloseForm}
              onSubmit={onSubmit}
            />
          </Suspense>
        )}
      </>
    );
  };

  return FormHOC;
};
