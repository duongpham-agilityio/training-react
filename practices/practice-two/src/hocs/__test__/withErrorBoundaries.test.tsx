// Mocks
import 'components/__test__/mocks/image.test';

import { render } from '@testing-library/react';

// Constants
import { MESSAGES } from '@constants';

// HOCs
import { withErrorBoundaries } from 'hocs/withErrorBoundaries';

const RejectComponent = () => {
  const isCondition = true;

  if (isCondition) throw new Error(MESSAGES.ERROR_DESCRIPTION);

  return <p>Reject</p>;
};

const ResolveComponent = () => {
  return <p>Resolve</p>;
};

describe('withErrorBoundaries', () => {
  it('Match to snapshot', () => {
    const Component = withErrorBoundaries(ResolveComponent);
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });

  it('Render reject component', () => {
    const Component = withErrorBoundaries(RejectComponent);
    const { getByText } = render(<Component />);

    expect(getByText(MESSAGES.ERROR_DESCRIPTION)).toBeInTheDocument();
  });
});
