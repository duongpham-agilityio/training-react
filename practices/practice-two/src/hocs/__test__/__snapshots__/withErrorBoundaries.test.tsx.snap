// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withErrorBoundaries Match to snapshot 1`] = `
<div>
  <p>
    Resolve
  </p>
</div>
`;
